### PR TITLE
mds: handle discontinuous mdsmap

### PR DESCRIPTION
**Pull request by ukernel**
There are two cases that a mds can get discontinuous mdsmap:
 - the mdsmap was sent by other mds
 - connection to monitor was reset

Monitor does not preserve old version mdsmap. There is no easy way to
ensure mds always gets continuous mdsmap. Instead, making mds handle
discontinuous mdsmap is not difficult. When failover happens in multimds
cluster, survivor mds may miss the mdsmap that indicates old mds failed
and/or the mdsmap that indicates new mds started to replay. But the
survivor mds always get the mdsmap that indicates the new mds entered
resolve state.

Fixes: http://tracker.ceph.com/issues/24856
Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

### CODE REVIEW

File: src/mds/MDSContext.cc [hunk 1]
```diff
@@ -28,6 +28,9 @@ void MDSInternalContextBase::complete(int r) {
   assert(mds != NULL);
   assert(mds->mds_lock.is_locked_by_me());
   MDSContext::complete(r);
+
+  // avoid unhealthy heartbeat map when finish_contexts() processes large context list
+  mds->heartbeat_reset();
```
Comment (batrick) [inline src/mds/MDSContext.cc]: This commit should be a separate PR.

Also, if we are processing a context list long enough to cause the heartbeat to timeout, then we are also not serving incoming requests. The context list should never get the long? I think we need to take a different approach.

File: src/mds/MDSRank.cc [hunk 2]
```diff
@@ -1740,6 +1737,61 @@ void MDSRankDispatcher::handle_mds_map(
   if (g_conf->mds_dump_cache_on_map)
     mdcache->dump_cache();
 
+  // mdsmap and oldmap can be discontinuous. failover might happen in the missing mdsmap.
+  set<mds_rank_t> restart;
+  if (state >= MDSMap::STATE_RESOLVE) {
```
Comment (batrick) [inline src/mds/MDSRank.cc]: I assume we don't need to mark the down MDSs when `< MDSMap::STATE_RESOLVE` because the MDS won't have any connections yet with the other ranks? Can you please add a comment?

**Comment by ukernel** [comment]
First stop mds.0, wait until it get marked laggy. then restart mds.1. when
mds.1 enters resolve state, resume execution of mds.0. .

I think you didn't stop mds.0 long enough, to get monc conntion reset.

File: src/mds/MDSContext.cc (hunk 1 above)
Comment (ukernel) [reply src/mds/MDSContext.cc]: The context list are delayed messages (caused by frozen subtree). how large it is depends on how many messages were sent by clients while subtree was frozen.

**Comment by batrick** [comment]
> First stop mds.0, wait until it get marked laggy. then restart mds.1. when mds.1 enters resolve state, resume execution of mds.0. . I think you didn't stop mds.0 long enough, to get monc conntion reset.

I must be missing a step because as soon as mds.0 is marked laggy it is eligble to be replaced by a standby. There are no standbys in this vstart cluster I'm testing with **but** as soon as mds.1 is restarted, it comes back as a standby and replaces mds.0. So:

````
2018-07-13 18:57:13.182 7f531ecb2700  1 -- 127.0.0.1:6847/149701165 --> 127.0.0.1:6827/24575 -- mgrreport(unknown.a +0-0 packed 214) v6 -- 0xfe220f2000 con 0
2018-07-13 18:57:36.089 7f5323cbc700  1 -- 127.0.0.1:6847/149701165 >> 127.0.0.1:6848/1715856436 conn(0xfe21e2c300 :6847 s=STATE_OPEN pgs=8 cs=1 l=0).read_bulk peer close file descriptor 147
2018-07-13 18:57:36.089 7f5323cbc700  1 -- 127.0.0.1:6847/149701165 >> 127.0.0.1:6848/1715856436 conn(0xfe21e2c300 :6847 s=STATE_OPEN pgs=8 cs=1 l=0).read_until read failed
2018-07-13 18:57:36.089 7f5323cbc700  1 -- 127.0.0.1:6847/149701165 >> 127.0.0.1:6848/1715856436 conn(0xfe21e2c300 :6847 s=STATE_OPEN pgs=8 cs=1 l=0).process read tag failed
2018-07-13 18:57:36.089 7f5322cba700  1 heartbeat_map is_healthy 'MDSRank' had timed out after 15
2018-07-13 18:57:36.089 7f5320cb6700  1 -- 127.0.0.1:6847/149701165 <== mon.2 127.0.0.1:40190/0 275 ==== mdsmap(e 83) v1 ==== 780+0+0 (3911091795 0 0) 0xfe21d04800 con 0xfe21db2e00
2018-07-13 18:57:36.089 7f5320cb6700  5 mds.a handle_mds_map epoch 83 from mon.2
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a      my compat compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,7=mds uses inline data,8=no anchor table,9=file layout v2,10=snaprealm v2}
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a  mdsmap compat compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,8=no anchor table,9=file layout v2,10=snaprealm v2}
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a map says I am 127.0.0.1:6847/149701165 mds.0.75 state up:active
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a handle_mds_map: handling map as rank 0
2018-07-13 18:57:36.089 7f5320cb6700  4 mds.0.75 set_osd_epoch_barrier: epoch=30
2018-07-13 18:57:36.089 7f5320cb6700  1 -- 127.0.0.1:6847/149701165 <== mon.2 127.0.0.1:40190/0 276 ==== osd_map(31..31 src has 1..31) v4 ==== 299+0+0 (1720088884 0 0) 0xfe21b91e40 con 0xfe21db2e00
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.0.snap check_osd_map need_to_purge={}
2018-07-13 18:57:36.089 7f5320cb6700  7 mds.0.server operator(): full = 0 epoch = 31
2018-07-13 18:57:36.089 7f5320cb6700  4 mds.0.75 handle_osd_map epoch 31, 1 new blacklist entries
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.0.server apply_blacklist: killed 0
2018-07-13 18:57:36.089 7f5320cb6700 10 monclient: _renew_subs
2018-07-13 18:57:36.089 7f5320cb6700 10 monclient: _send_mon_message to mon.c at 127.0.0.1:40190/0
2018-07-13 18:57:36.089 7f5320cb6700  1 -- 127.0.0.1:6847/149701165 --> 127.0.0.1:40190/0 -- mon_subscribe({osdmap=32}) v3 -- 0xfe21e066c0 con 0
2018-07-13 18:57:36.089 7f5320cb6700  1 -- 127.0.0.1:6847/149701165 <== mon.2 127.0.0.1:40190/0 277 ==== mdsmap(e 84) v1 ==== 657+0+0 (24463606 0 0) 0xfe21d04a80 con 0xfe21db2e00
2018-07-13 18:57:36.089 7f5320cb6700  5 mds.a handle_mds_map epoch 84 from mon.2
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a      my compat compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,7=mds uses inline data,8=no anchor table,9=file layout v2,10=snaprealm v2}
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a  mdsmap compat compat={},rocompat={},incompat={1=base v0.20,2=client writeable ranges,3=default file layouts on dirs,4=dir inode in separate object,5=mds uses versioned encoding,6=dirfrag is stored in omap,8=no anchor table,9=file layout v2,10=snaprealm v2}
2018-07-13 18:57:36.089 7f5320cb6700 10 mds.a  peer mds gid 7038 removed from map
````

Zheng, I need you to come up with a reproducer and test. Also, it's not clear from the tracker ticket or this PR/commit what the consequence of a discontinous map is. I've been assuming that this will reproduce the MDS stuck in resolve but maybe not? Please amend the ticket and commit message to say what the real problem we're fixing is.

**Comment by ukernel** [comment]
need to assign MDSs to specified  ranks  (mds_standby_for_rank config)

**Comment by batrick** [comment]
> need to assign MDSs to specified ranks (mds_standby_for_rank config)

Zheng, I don't have time to keep testing this even if I'd like to and this is time-sensitive. I need you to come up with a QA test.

**Comment by ukernel** [comment]
@batrick qa done

**Comment by ukernel** [comment]
For luminous, https://github.com/ukernel/ceph/commits/luminous-24856.  there is a major difference in qa part

**Comment by batrick** [comment]
@ukernel, the `vstart_runner.py` fixes are in master. Please rebase.
